third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -0,0 +1,89 @@
+#include "mlir/Dialect/LLVMIR/ROCDLDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+
+#define GEN_PASS_CLASSES
+#include "TritonAMDGPUTransforms/Passes.h"
+
+using namespace mlir;
+namespace ttg = mlir::triton::gpu;
+namespace tt = mlir::triton;
+
+class Pingponger {
+  scf::ForOp forOp;
+  SmallVector<Operation *> gLoadOps;
+  SmallVector<Operation *> lLoadOps;
+  SmallVector<Operation *> dotOps;
+
+public:
+  Pingponger(scf::ForOp forOp) : forOp(forOp) {};
+  void getDotPingponged();
+};
+
+void Pingponger::getDotPingponged() {
+
+  OpBuilder builder(forOp);
+  MLIRContext *ctx = forOp.getContext();
+  Location loc = forOp.getLoc();
+  IntegerAttr schedMaskAttr = IntegerAttr::get(IntegerType::get(ctx, 32), 6);
+  IntegerAttr schedMaskAttr0 = IntegerAttr::get(IntegerType::get(ctx, 32), 0);
+  IntegerAttr lowPrioAttr = IntegerAttr::get(IntegerType::get(ctx, 16), 0);
+  IntegerAttr highPrioAttr = IntegerAttr::get(IntegerType::get(ctx, 16), 1);
+  auto f16_ty = builder.getF16Type();
+
+  forOp->walk([&](triton::LoadOp op) { gLoadOps.push_back(op); });
+  forOp->walk([&](ttg::LocalLoadOp op) { lLoadOps.push_back(op); });
+  forOp->walk([&](triton::DotOp op) { dotOps.push_back(op); });
+
+  if (gLoadOps.size() != 2 || lLoadOps.size() != 2 || dotOps.size() != 1)
+    return;
+
+  // Splitting loading A and B inorder to prevent global/local load units
+  // from the congestion.
+  // Locate global load at the end. Otherwise, local_load at the end of
+  // the sequence will be overlapped with the first local_stores from
+  // the other warp.
+  // sched.barriers to keep the order.
+  lLoadOps[0]->moveBefore(gLoadOps[0]);
```

Comment:
This is illegal if the loop was not pipelined.

Reply:
Fixed in compiler.py